capsules/src/ieee802154/driver.rs

```diff
@@ -0,0 +1,827 @@
+    /// `None`.
+    fn lookup_key(&self, level: SecurityLevel, key_id: KeyId) -> Option<([u8; 16])> {
+        self.keys.and_then(|keys| {
+            keys[..self.num_keys.get()]
+                .iter()
+                .find(|key| key.level == level && key.key_id == key_id)
+                .map(|key| key.key)
+        })
+    }
+}
+
+impl<'a> Driver for RadioDriver<'a> {
+    /// Setup buffers to read/write from.
+    ///
+    /// ### `allow_num`
+    ///
+    /// - `0`: Read buffer. Will contain the received frame.
+    /// - `1`: Write buffer. Contains the frame payload to be transmitted.
+    /// - `2`: Config buffer. Used to contain miscellaneous data associated with
+    ///        some commands because the system call parameters / return codes are
+    ///        not enough to convey the desired information.
+    fn allow(&self, appid: AppId, allow_num: usize, slice: AppSlice<Shared, u8>) -> ReturnCode {
+        match allow_num {
+            0 | 1 | 2 => {
+                self.do_with_app(appid, |app| {
+                    match allow_num {
+                        0 => app.app_read = Some(slice),
+                        1 => app.app_write = Some(slice),
+                        2 => app.app_cfg = Some(slice),
+                        _ => {}
+                    }
+                    ReturnCode::SUCCESS
+                })
+            }
+            _ => ReturnCode::ENOSUPPORT,
+        }
+    }
+
+    /// Setup callbacks.
+    ///
+    /// ### `subscribe_num`
+    ///
+    /// - `0`: Setup callback for when frame is received.
+    /// - `1`: Setup callback for when frame is transmitted.
+    fn subscribe(&self, subscribe_num: usize, callback: Callback) -> ReturnCode {
+        match subscribe_num {
+            0 => {
+                self.do_with_app(callback.app_id(), |app| {
+                    app.rx_callback = Some(callback);
+                    ReturnCode::SUCCESS
+                })
+            }
+            1 => {
+                self.do_with_app(callback.app_id(), |app| {
+                    app.tx_callback = Some(callback);
+                    ReturnCode::SUCCESS
+                })
+            }
+            _ => ReturnCode::ENOSUPPORT,
+        }
+    }
+
+    /// IEEE 802.15.4 MAC device control.
+    ///
+    /// For some of the below commands, one 32-bit argument is not enough to
+    /// contain the desired input parameters or output data. For those commands,
+    /// the config slice `app_cfg` is used as a channel to shuffle information
+    /// between kernel space and user space. The expected size of the slice
+    /// varies by command, and acts essentially like a custom FFI. That is, the
+    /// userspace library MUST `allow()` a buffer of the correct size, otherwise
+    /// the call is EINVAL. When used, the expected format is described below.
+    ///
+    /// ### `command_num`
+    ///
+    /// - `0`: Return radio status. SUCCESS/EOFF = on/off.
```

Comment:
Tock has a convention that calling command 0 should always return >= 0 if the driver is present (i.e. the kernel has support for it) and < 0 if the kernel isn't compiled with that `Driver`. So here, you can either define a return of `1` as on and `0` as off (so that whether the radio is on or off userland can still detect that there is a 15.4 driver present) or you can move each command up a number and add a new command 0 that just returns `SUCCESS` when called (as many other capsule drivers do).